File: rules/commandinjection.js
const fs = require('fs');
const vscode = require('vscode');

var smell = {

    detect : (fileName, token) => {
        try{
            if(token.hasOwnProperty("name")) var name= token.name;
            if(token.hasOwnProperty("line")) var lineno = token.line;
            if(token.hasOwnProperty("type")) var tokenType = token.type;
            
            const MSG = 'possible presence of command injection'
            
            const WARNING_MSG = MSG+' at line '+ lineno;
            const WARNING_MSG_ON_RETURN = token.hasOwnProperty("returnLine") ? WARNING_MSG+ token.returnLine : null;
    
            const shellFunctions = ['sys.argv', 'subprocess.Popen', 'os.system', 'os.popen','subprocess.run','popen2.Popen4',
                                    'argparse.ArgumentParser','getopt.getopt', 'os.execle','os.execl', 'popen2.Popen3'
                                ];
            
            if(tokenType == "variable" && token.hasOwnProperty("valueSrc")) {
                if(shellFunctions.includes(token.valueSrc) || smell.isExtendedShellFunction(token.valueSrc)) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                // vscode.commands.executeCommand('revealLine',{'lineNumber':lineno, 'at':'top'});
            }
            else if(tokenType == "function_call" && (shellFunctions.includes(name) || smell.isExtendedShellFunction(name))) smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
            else if (tokenType == "function_def" && token.hasOwnProperty('return') && token.return != null) {
                for(const funcReturn of token.return){
                    if(shellFunctions.includes(funcReturn) || smell.isExtendedShellFunction(funcReturn)) 
                    smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG_ON_RETURN);
                }
            }
        } catch(error){
            console.log(error);
        }
    },

    isExtendedShellFunction: functionName => {
        const shellFunctions = ['sys.argv','subprocess.Popen','os.system','os.popen','subprocess.run','argparse.ArgumentParser','getopt.getopt']
        
        for (const name of shellFunctions) {
            if(functionName.includes(name)) return true;
        }

        return false;
    },

    triggerAlarm: (fileName, MSG, lineno, WARNING_MSG) => {
        let backslashSplittedFilePathLength = fileName.split("/").length
        let filenameFromPath = fileName.split("/")[backslashSplittedFilePathLength - 1]

        vscode.window.showWarningMessage(MSG +" : "+ filenameFromPath+":"+lineno);
        console.log( "\u001b[1;31m"+"warning: "+MSG +"  location:"+ fileName+":"+lineno);
        fs.appendFileSync(__dirname+'/../warning-logs/project_warnings.csv', fileName+" ,"+WARNING_MSG+"\n");
        
        // fs.appendFile(__dirname+'/../logs/project_warnings.csv', fileName+","+WARNING_MSG+"\n", (err) => err ? console.log(err): "");
    }
}

module.exports = smell;

File: py-scripts/filters/commandinjection.py
import re

class CommandInjection:
    '''This is the class for detecting command injection in code'''

    def __init__(self):
        self.shell_methods = [
                        'sys.argv', 'subprocess.Popen', 'os.system', 'os.popen','subprocess.run','popen2.Popen4',
                        'argparse.ArgumentParser','getopt.getopt', 'os.execle', 'os.execl','popen2.Popen3'
            ]
            
        self.warning_message = 'command injection'

    def detect_smell(self, token, src_file):
        try:
            if token.__contains__("line"): lineno = token["line"] 
            if token.__contains__("type"): tokenType = token["type"]
            if token.__contains__("name"): name = token["name"]
            if token.__contains__("args"): args = token["args"]
            if token.__contains__("hasInputs"): containsUserInput =  token["hasInputs"]
            
            if tokenType == "variable" and token.__contains__("valueSrc") and token["valueSrc"] is not None and token["valueSrc"] != 'initialization':
                if token["valueSrc"].strip() in self.shell_methods or self.is_extended_shell_command_names(token["valueSrc"].strip()):
                    self.trigger_alarm(src_file, lineno)
            
            elif tokenType == "function_call" and name is not None and (name.strip() in self.shell_methods or self.is_extended_shell_command_names(name.strip())): 
                self.trigger_alarm(src_file, lineno)
        
            elif tokenType == "function_def" and token.__contains__('return') and token["return"] is not None: 
                for func_return in token['return']:
                    if isinstance(func_return, str) and (func_return in self.shell_methods or self.is_extended_shell_command_names(func_return)):
                        self.trigger_alarm(src_file, lineno)
            
        except Exception as error: 
            print("command injection error"+str(error))
        
    def is_extended_shell_command_names(self, method_name):
        
        if isinstance(method_name, str) is False: return False
        elif len(method_name) == 0: return False

        self.shell_methods = ['sys.argv', 'subprocess.Popen', 'os.system', 'os.popen','subprocess.run', 'argparse.ArgumentParser', 'getopt.getopt']
        
        for name in self.shell_methods:
            if name in method_name: 
                return True

        return False

    
    def trigger_alarm(self, src_file, lineno):
        print(src_file +":"+ str(lineno)+" ,"+self.warning_message)
    